Loadpoint UI: add priority

diddip21 opened the pull request:
fixes #19650 

- [x] aktuelle Priorität wird hinter dem Loadpoint Titel als Icon angezeigt
- [x] Sind keine Prios gesetzt (default 0) wird kein Icon angezeigt
- [x] Klick auf Icon kann direkt die Priorität verändern (natives Select, Screenshot)
- [x] Wir verwenden die Scale (-3 bis +3) mit übersetzten Namen (..., low, normal, high, ...)
- [ ] Ladepunkt Layout im LG Breakpoint umstellen (siehe https://github.com/evcc-io/evcc/issues/19650#issuecomment-2778365036)

**Kompatibilität/Große Installationen**
- [x] Werte außerhalb der reduzierten Skala sind weiter möglich.
- [x] Priorität kann auch über den Ladepunkt-Settings-Dialog (https://github.com/evcc-io/evcc/issues/19650#issuecomment-2765276642) eingestellt werden. Darstellung als Dropdown mit Werten von -10 bis +10.
- [x] Priorität-Auswahl im Config-UI Ladepunkt Dialog wird entfernt (redundant).

![grafik](https://github.com/user-attachments/assets/f24ad5f4-1b73-4405-8cd8-9aeba507d573)
![grafik](https://github.com/user-attachments/assets/497fd8ec-8870-47ff-957f-c93bf69c0ca6)
![grafik](https://github.com/user-attachments/assets/f4c4a9a1-0627-44fb-90d4-4021d921c408)
![image](https://github.com/user-attachments/assets/1d44d38e-2fae-480e-b6df-59a74b1ef7bd)

Commit:
fix double naming e.g. Buderus Buderus (SG Ready) in docs
added description link to emsesp

Commit:
added Loadpoint PrioButton

Commit:
added missing zero

Commit:
fix Breakpoint at Title
added PrioButton
 - effectivePriority used -> disableEdit
 - priority set -> show PrioButton

Commit:
fix missing toml entries

Commit:
fix allow negative and positiv priority via api

Commit:
fix camelCase
fix toml sort

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

andig commented:
Es gibt keine Very Low... very High Skala. Prio ist rein numerisch.

diddip21 commented:
Der Dropdown ist rein zur Übersicht very Low bis very High und nur in der UI.
Als Value wird immer ein numerischer Wert verwendet

andig commented:
Die Prio ist aktuell numerisch. Dem jetzt andere Bezeichnungen zu geben finde ich verwirrend. Die Icons nutzen wir bereits für Phasenumschaltung und andere Mechanismen- die Doppelung wäre für mich ebenfalls verwirrend. Wie wäre es denn mit P1..10 stattdessen? Sinnvoll wäre auch anzuzeigen, ob ein Ladepunkt niedrigerer Priorität gerade übersteuert wird?

diddip21 commented:
Die beiden offenen Todos beziehen sich noch darauf, die Settings für die Ladepunkt Prio aus den Ladepunkt Settings zu entfernen und in den Ladepunkt Dialog zu bringen.
Die schnell Einstellungen mit Low und High sind zum direkten umstellen und aus meiner Sicht unmissverständlich im Gegensatz zu Prio 0 bis 10 (Ist jetzt Prio1 höherwertiger als Prio10 oder andersrum) mit low/high ist das eindeutig.
> Sinnvoll wäre auch anzuzeigen, ob ein Ladepunkt niedrigerer Priorität gerade übersteuert wird?

Was meinst du damit ? Die übersteuerung aus dem Fahrzeug ist berücksichtigt über die effectivePriority

naltatis commented on templates/definition/charger/emsesp.yaml:
```diff
@@ -2,15 +2,18 @@ template: emsesp
   - brand: Buderus
     description:
-      generic: Buderus (SG Ready)
+      generic: SG Ready
```
unrelated

VolkerK62 commented:
>Prio 0 bis 10 (Ist jetzt Prio1 höherwertiger als Prio10 oder andersrum) mit low/high ist das eindeutig.

Da ist was dran. Gab schon mehrfach Diskussionen, wo Nutzer der Meinung waren Prio1 wäre die höchste Priorität.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+		type="button"
+		class="btn btn-sm btn-outline-secondary position-relative border-0 pt-0"
+		data-testid="loadpoint-prio-button"
+		:class="{ 'opacity-25': !editable }"
```
Unnötige Redundanz. Hier wissen wir doch schon , dass das nicht editable ist. Gleiches gilt auch für Tooltip und den anderen Button.

naltatis replied:
Semantisch ist das ein Disabled Button (Interaktionsfläche, die gerade nicht verwendbar ist). Vmtl. kannst du dir dann auch das manuelle opacity sparen.

diddip21 replied:
Redundanzen hab ich entfernt im näcshten commit.
mit dem Disabled habe ich gespielt, da ist das Problem dass dann kein Tooltip hochkommt, müsste dann noch ein anderes Element drum herum bauen für den Tooltip.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+		data-bs-html="true"
+		:title="tooltipTitle"
+		:aria-label="editable ? $t('main.loadpointPrio.prioTooltip') : null"
+	>
```
Select Element in Button ist ungültiges Markup. Hier brauchen wir den Button eigentlich gar nicht. Müssen nur schauen, dass sich das Styling/Touchtarget Button-Style verhält.

naltatis replied:
Schau dir mal die Implementierung von `CustomSelect` an. Der Slot direkt direkt in ein `label` gerendert und ist damit bereits eine Interaktionsfläche.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+		type="button"
+		class="btn btn-sm btn-outline-secondary position-relative border-0 pt-0"
+		data-testid="loadpoint-prio-button"
+		:class="{ 'opacity-25': !editable }"
+		data-bs-toggle="tooltip"
+		data-bs-html="true"
+		:title="tooltipTitle"
+		:aria-label="editable ? $t('main.loadpointPrio.prioTooltip') : null"
```
```suggestion
```

diddip21 replied:
Der Tooltip sollte meiner Meinung nach auch im editierbaren modus erscheinen, da dieser schnell und kompakt die den echten numerischen Wert anzeigt (falls über die Icons nicht eindeutig)

naltatis replied:
> kompakt die den echten numerischen Wert anzeigt (falls über die Icons nicht eindeutig)

Wenn das wirklich ein Problem ist (bin ich noch nicht von überzeugt) müssen wir vielleicht doch noch mal über die Icon-Sprache nachdenken.

Ich bin auch, jetzt wo man es konkret im Code und in der Anwendung sieht, noch nicht so glücklich damit, dass es einmal nur eine Anzeige und einmal Veränderbar ist. Liegt an unserer ja doch etwas komplizierteren Logik (Fahrzeug kann Ladepunkt übersteuern), die ich bei der ersten Idee noch nicht so prominent auf dem Zettel hatte.

Vielleicht ist es doch einfacher, wenn die Prio nicht in-place editierbar ist, sondern wir direkt die Loadpoint-Settings (wie in deinem ersten Vorschlag) beim Klick öffnen.

Kommt der Wert vom Fahrzeug könnten wir beim Klick die Fahrzeug-Settings öffnen. Fahrzeug-Settings gibts (noch) nicht als eigenen Dialog. De-factor ist aber der "Ankunft"-Reiter im Ladeplan Dialog (wenn es ein Fahrzeug gibt) der Ort an dem wir schon heute Fahrzeug-Einstellungen editierbar (bspw. minsoc) haben. Hier könnten wir auch ein Prio-Setting hinzufügen (analog Ladepunkt). Dann könnte auch diese Einstellung im Config-UI (Fahrzeuge) entfernt werden. Wäre aus Symmetrie und Konsistenzgründe sicher gut.

diddip21 replied:
Tooltip mit der Prio als Wert ist entfernt, es kommt nur noch der Hinweis dass es vom Fahrzeug übersteuert wird.
die Quick Priorisierung ist auch noch aktiv, sollen wir diese tatsächlich entfernen? per Loadpoint Config lässt diese sich nun im Dropdown style ändern.

naltatis replied:
So, bin gerade mal die PR bei mir durchgegangen und hab einige kleinere Anpassungen direkt comittet. Ja, lass uns das Dropdown hier rausnehmen. Dann haben wir nicht mehr zwei Skalen (+3...-3) vs. die größere in den Settings. Lass uns Prio auch gleich in den Ankunft-Reiter bei bekannten Fahrzeugen mit aufnehmen.

Tooltip: Würde ich jetzt simple halten. Kein HTML, einfach "Priorität: +2". Vielleicht noch mit Ergänzung falls vom Fahrzeug übersteuert wurde: "Priorität: +2 (Fahrzeug)".

Den 0-Wert hab ich jetzt rausgenommen und auch die Logik, dass wir immer anzeigen oder gar nicht habe ich wieder reverted. Das ist sinnvoll wenn es ein direkt steuerbares Element ist, aber wenn das über die Settings gemacht wird können wir die 0-Zustand auch einfach weglassen.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+import _minus3 from "./minus-3.vue";
+import _plus1 from "./plus-1.vue";
+import _plus2 from "./plus-2.vue";
+import _plus3 from "./plus-3.vue";
```
Hier sollten wir eine eigene Icon-Komponente machen, die die Logik kapselt. Analog zum `BatteryIcon`. Ich denke an sowas: `<PriorityIcon :value="-2" />`.  Die kann dann auch die Endgecases (`-4`) handeln.

andig replied:
Ich weise nochmal darauf hin, dass Prio ein numerisches Konzept ist. Wir haben anscheinend noch keine gemeinsame Sicht.

diddip21 replied:
https://github.com/evcc-io/evcc/pull/20499#issuecomment-2791945291
Die Priorität im Hintergrund numerisch zu halten, da sind wir uns alle einig, denke ich.
Für den einfachen Endanwender ist eine namentliche Darstellung unmissverständlich.
Über die Loadpoint Config wäre die Darstellung folgendermaßen (letztes Bild) https://github.com/evcc-io/evcc/issues/19650#issuecomment-2765276642 und auch hier durchaus unmissverständlich.
Was meinst du dazu @andig ?

diddip21 replied:
![image](https://github.com/user-attachments/assets/53537e0c-280a-49ad-b05c-3720437a65b9)
Hier nochmal direkt die mögliche Darstellung

naltatis replied:
Lass uns hier einfach das Dropdown aus Config UI nehmen.

naltatis replied:
Ist einfach zu bedienen, technisch simpler (alte/exotische Browser Support).

naltatis commented on assets/js/components/Loadpoints/Loadpoint.vue:
```diff
@@ -11,14 +11,24 @@
 					<div class="text-truncate">
 						{{ loadpointTitle }}
 					</div>
+					<PrioButton
```
Naming: eher `PrioritySelect` (analog `Limit[Soc|Energy]Select`)

naltatis commented on assets/js/components/Loadpoints/Loadpoint.vue:
```diff
@@ -11,14 +11,24 @@
+								: priority
+						"
+						:editable="effectivePriority === priority"
+						@update:prio="setPriority"
```
Naming: siehe `LimitSocSelect`

naltatis commented on assets/js/components/Loadpoints/Loadpoint.vue:
```diff
@@ -11,14 +11,24 @@
+								? effectivePriority
+								: priority
+						"
+						:editable="effectivePriority === priority"
```
zu viel (redundante) Logik im Template. Ich würd `priority` und `effectivePriority` als Props in die Komponente geben. Die muss dann selbst die richtigen Werte ableiten.

Ob die Komponente angezeigt wird gehört weiterhin hier in den Loadpoint. Hier würde ich die Regel machen, dass wir das nicht pro Loadpoint entscheiden sondern bereit auf der Ebene darüber (Loadpoints). Die gibt dann ein `showPriority` an alle Loadpoint Komponenten. Logik wäre: Sobald ein Loadpoint einen nicht `0` Priorität hat wird das Element an allen Loadpoints angezeigt.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+			if (!this.editable) {
+				title += `<div class="mt-1">${this.$t("main.loadpointPrio.prioTooltip")}</div>`;
+			}
+			return `<div class="text-start">${title}</div>`;
```
Tooltip nur, wenns nicht editierbar ist. Auf Mobile funktionieren Tooltips ohnehin nicht sehr gut.
Den aktuellen Wert würde ich nicht ins Tooltip packen. Lass uns den simpel halten und nur im "Durch Fahrzeug übersteuert" Fall als reinen Texttooltip nutzen.

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+				{ value: 0, name: this.$t("main.loadpointPrio.neutral") },
+				{ value: 1, name: this.$t("main.loadpointPrio.slightlyHigh") },
+				{ value: 2, name: this.$t("main.loadpointPrio.high") },
+				{ value: 3, name: this.$t("main.loadpointPrio.veryHigh") },
```
Ich find die "Slightly" Stufen ungewöhnlich. Ist lustigerweise der erste ChatGPT Vorschlag gewesen. Lass uns hier auf kurz und symmetrisch gehen:

```
sehr hoch
höher
hoch
normal
niedrig
niedriger
sehr niedrig
```

bzw.

```
very high
higher
high
normal
low
lower
very low
```

Liegt der aktuelle Wert außerhalb der Skala (-7, 10) könnten wir ihn als zusätzlichen Eintrag mit in die Optionen aufnehmen (`even lower [-7]`). Dann haben wir Immer einen ausgewählten Zustand. Das gleiche machen wir aktuell bei bspw. beim Current Limit und den Auswahldialogen im Battery-Dialog (bufferSoc, ...).

naltatis commented on assets/js/components/Loadpoints/PrioButton/PrioButton.vue:
```diff
@@ -0,0 +1,144 @@
+	},
+	props: {
+		prio: { type: Number },
+		size: { type: String, default: "sm" },
```
Brauchen wir aktuell nicht.

naltatis commented on core/loadpoint_api.go:
```diff
@@ -219,7 +219,7 @@ func (lp *Loadpoint) SetPriority(prio int) {
 	lp.log.DEBUG.Println("set priority:", prio)
 
-	if lp.Priority != prio {
+	if lp.Priority != prio || prio == 0 {
```
Warum ist das erforderlich?

diddip21 replied:
Hab gerade nochmal geprüft, er nimmt sonst über die API keinen 0 Wert an.

naltatis replied:
Aber wieso sollte das so sein? Kann ja eigentlich nur passieren, wenn `lp.Priority` bereits 0 ist.

diddip21 replied:
wir in #20795 behandelt

naltatis commented:
@diddip21 Mega! Der PR sieht schon richtig gut aus. 🙌

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

Commit:
added Prio to LoadpointModalSettings

Commit:
remove Priority from Loadpoint Config

Commit:
Merge branch 'feat-ui/loadpoint-prio' of https://github.com/diddip21/evcc into feat-ui/loadpoint-prio

Commit:
fix some errors

Commit:
removed redundancy
added low / high text to slider

naltatis committed:
Merge branch 'master' into feat-ui/loadpoint-prio

andig commented:
> @andig wir müssen `EffectivePriority` jedes mal neu publishen, wenn `setPriority` aufgerufen wird. 

Das wäre sinnvoll. Dafür brauchts ein lock-freies EffectivePriority

> Aktuell passiert das nur initial und beim Fahrzeugwechsel. Gute Idee wie wir das machen? Es in die Funktion aufzunehmen führt zu nem Lock. Alternativ muss das in die Regelschleife und wir publishen einfach immer.

Das passiert schon?

Commit:
Merge branch 'feat-ui/loadpoint-prio' of https://github.com/diddip21/evcc into feat-ui/loadpoint-prio

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

Commit:
changed PriorityButton to PrioritySelect
created separated PriorityIcon from PrioritySelect
added showPriorityOnAllLoadpoints (shows priority if a priority is set somewhere)

Commit:
breakpoint LG grid size

diddip21 committed:
Merge branch 'master' into feat-ui/loadpoint-prio

naltatis committed:
lg loadpoint layout; icons for extreme prios; clearer dropdown values [-10..+10]

naltatis committed:
Merge branch 'master' into feat-ui/loadpoint-prio

naltatis commented:
Hier nochmal ein paar Screenshots der Prio Zustände am Ladepunkt.

**0 (default), -1**

<img width="1085" alt="Bildschirmfoto 2025-04-24 um 17 42 01" src="https://github.com/user-attachments/assets/f68c46d9-2935-4d60-8560-e611c4465c73" />

---

**extreme Werte (Zahl + Richtung bei größer 3)**

<img width="1025" alt="Bildschirmfoto 2025-04-24 um 17 40 57" src="https://github.com/user-attachments/assets/96141729-c2c1-443c-8d92-9d5cf5caa0c5" />

<img width="1098" alt="Bildschirmfoto 2025-04-24 um 17 41 41" src="https://github.com/user-attachments/assets/c6319b1d-cfdb-40c7-bd58-58cc5be2d761" />

---

**Ladepunkt Einstellungen**

<img width="958" alt="Bildschirmfoto 2025-04-24 um 17 41 18" src="https://github.com/user-attachments/assets/a6184e3f-6c89-40cf-b7f5-fcced5c6cdbd" />

<img width="658" alt="Bildschirmfoto 2025-04-24 um 17 41 12" src="https://github.com/user-attachments/assets/9c3240aa-034f-49e4-807e-6147b7255f67" />

andig commented:
Was bedeuten die hoch/runter Pfeile über der Zahl?
Müssen wir wirklich negative Prioritäten einführen? Das wäre schön zu vermeiden. Der Einmalaufwand z.B. 3 Ladepunkten eine "höhere" Prio zu geben damit der 4. auf 0 bleiben kann ist beherrschbar und kein Showstopper.

naltatis commented:
> Müssen wir wirklich negative Prioritäten einführen? Das wäre schön zu vermeiden.

Doch, finde ich aus Nutzersicht wichtig. Ich hab hier vier Ladepunkte. Einer ist ein Heizstab. Der ist in meinem Kopf "low prio". Mir gehts hier nicht um dein einmaligen Config Aufwand die anderen höher zu stellen sondern darum, dass wir "Mental Model" besser abbilden. Default = "mittlere Prio" und abhängig vom Anwendungsfall kann hoch oder runterpriorisiert werden.

> Die hoch/runter Pfeile über der Zahl sind nicht sofort offensichtlich.

Wie die anderen Status-Icons auch wird hier nen Tooltip geben der erklärt. Ich find das mit der Zahl optisch schon sehr stark und unruhig. Daher würde ich für die gängigen Anwendungsfälle (wenige Ladepunkte) gerne eine dezentere Darstellung (ohne Zahl) haben. Hier ist mir Ruhe wichtiger als Konsistenz mit den "Extremen". Den Punkt mit der Phasenumschaltung sehe ich allerdings. Dort nutzen wir immer einen Doppelpfeil (hoch/runter). Hier mal ein Screenshot. 

<img width="555" alt="Bildschirmfoto 2025-04-24 um 19 14 02" src="https://github.com/user-attachments/assets/d45cd6c7-60e2-4023-9973-1fc07b0de3c9" />

Vielleicht ist ein guter Kompromiss bei den Prios nur bei einem einfachen Pfeil zu bleiben und die Zahl bei der `+1`/`-1` wegzulassen, darüber dann aber mit der Kombination Zahl+Pfeil zu arbeiten.

> Würde es nicht sogar reichen zu zeigen, ob ein LP aufgrund seiner Prio gerade aktiv oder inaktiv wird?

Ich fänd das gut, wenn wir das sinnvoll immer dann anzeigen wenn es für den Nutzer relevant ist. Mir fällt hier aber keine wirklich gute Logik ein an der wir das festmachen. Können wir aber auch in einer späteren Iteration immer noch "intelligenter" machen.

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

TobiasHuber1980 commented:
> Müssen wir wirklich negative Prioritäten einführen? Das wäre schön zu vermeiden. Der Einmalaufwand z.B. 3 Ladepunkten eine "höhere" Prio zu geben damit der 4. auf 0 bleiben kann ist beherrschbar und kein Showstopper. Das wäre dann 0 (lowest) bis 10 (highest).
> 
Ich kann mich der Meinug von @andig nur anschliessen, ich finde -10 bis +10 einfach "too much". Logischer für mich wäre 0=default; 10 = highest.

Commit:
Merge branch 'master' into feat-ui/loadpoint-prio

Commit:
Merge branch 'feat-ui/loadpoint-prio' of https://github.com/diddip21/evcc into feat-ui/loadpoint-prio

diddip21 committed:
Merge branch 'master' into feat-ui/loadpoint-prio

Commit:
fix typescript error

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

diddip21 committed:
Merge branch 'evcc-io:master' into feat-ui/loadpoint-prio

realwuffi commented:
Bei mir in der Arbeit gabs auch immer Missverständnisse bzgl Prio und numerischen Werten, ich habe dann mal die Begriffe 'Important' und 'Not Important' dafür eingeführt und seitdem war alles klar.
Den Zahlenwert in der UI mit darzustellen muss eigentlich nicht sein. Und der Einfachheit halber sollten wohl 3 Werte (Neutral, Wichtig und Unwichtig) auch reichen, was dann ggfs auch einfacher mit zwei Symbolen (Neutral hat keins) angezeigt werden könnte.

Vielen Dank für Eure super Arbeit hier! Ich frei mich schon, wenn meine Wärmepumpe und der Heizstab hoffentlich diesen Herbst noch Realität werden :-)
